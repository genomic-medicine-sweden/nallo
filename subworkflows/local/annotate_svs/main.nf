include { SVDB_QUERY                         } from '../../../modules/nf-core/svdb/query/main'
include { ENSEMBLVEP_VEP as ENSEMBLVEP_SV    } from '../../../modules/nf-core/ensemblvep/vep/main'
include { TABIX_TABIX as TABIX_ENSEMBLVEP_SV } from '../../../modules/nf-core/tabix/tabix/main'

workflow ANNOTATE_SVS {

    take:
    ch_vcf                // channel: [mandatory] [ val(meta), path(vcf) ]
    ch_fasta              // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_sv_dbs             // channel: [mandatory] [ val(meta), path(csv) ]
    ch_vep_cache          // channel: [mandatory] [ path(cache) ]
    val_vep_cache_version //  string: [mandatory] default: 110
    ch_vep_extra_files    // channel: [mandatory] [ path(files) ]

    main:
    ch_versions = Channel.empty()
    ch_versions.view()
    ch_sv_dbs
        .map { meta, csv -> csv }
        .splitCsv ( header:true )
        .multiMap { row ->
            vcf_dbs:  row.filename
            in_frqs:  row.in_freq_info_key
            in_occs:  row.in_allele_count_info_key
            out_frqs: row.out_freq_info_key
            out_occs: row.out_allele_count_info_key
        }
        .set { ch_svdb_in }

    // Annotate with SVDB VCF "databases"
    SVDB_QUERY (
        ch_vcf,
        ch_svdb_in.in_occs.toList(),
        ch_svdb_in.in_frqs.toList(),
        ch_svdb_in.out_occs.toList(),
        ch_svdb_in.out_frqs.toList(),
        ch_svdb_in.vcf_dbs.toList(),
        []
        )
        ch_versions = ch_versions.mix(SVDB_QUERY.out.versions)

    ENSEMBLVEP_SV (
        SVDB_QUERY.out.vcf.map { meta, vcf -> [ meta, vcf, [] ] },
        "GRCh38",
        "homo_sapiens",
        val_vep_cache_version,
        ch_vep_cache,
        ch_fasta,
        ch_vep_extra_files
    )
    ch_versions = ch_versions.mix(ENSEMBLVEP_SV.out.versions)

    TABIX_ENSEMBLVEP_SV (
        ENSEMBLVEP_SV.out.vcf
    )
    ch_versions = ch_versions.mix(TABIX_ENSEMBLVEP_SV.out.versions)

    emit:
    vcf      = ENSEMBLVEP_SV.out.vcf       // channel: [ val(meta), path(vcf) ]
    tbi      = TABIX_ENSEMBLVEP_SV.out.tbi // channel: [ val(meta), path(tbi) ]
    versions = ch_versions                 // channel: [ path(versions.yml) ]
}
