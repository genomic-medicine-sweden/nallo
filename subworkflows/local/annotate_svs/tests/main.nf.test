nextflow_workflow {

    name "Test Workflow ANNOTATE_SVS"
    script "../"
    workflow "ANNOTATE_SVS"
    config "./nextflow.config"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'hg38' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }
        }

        run("SEVERUS") {
            script "../../../../modules/nf-core/severus/main"
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, project: 'project' ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam.bai', checkIfExists: true),
                    [],
                    [],
                    []
                ]
                input[1] = [[],[]]
                """
            }
        }

        run("UNTAR") {
            script "../../../../modules/nf-core/untar/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'vep_cache' ],
                    file(params.pipelines_testdata_base_path + 'reference/vep_cache_test_data.tar.gz', checkIfExists:true)
                ]
                """
            }
        }

    }

    test("severus, fasta, svdb, vep_cache, '110', vep_plugins") {

        when {
            workflow {
                """
                input[0] = SEVERUS.out.all_vcf
                input[1] = GUNZIP.out.gunzip
                input[2] = Channel.of([
                    [ id: 'svdb_dbs' ],
                    file(params.pipelines_testdata_base_path + 'testdata/svdb_dbs.csv', checkIfExists: true)
                ])
                input[3] = UNTAR.out.untar.map { meta, cache -> cache }
                input[4] = Channel.value('110')
                input[5] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    path(workflow.out.vcf.get(0).get(1)).vcf.variantsMD5,
                    file(workflow.out.tbi.get(0).get(1)).name
                ).match() }
            )
        }

    }

    test("severus, fasta, svdb, vep_cache, '110', vep_plugins -stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = SEVERUS.out.all_vcf
                input[1] = GUNZIP.out.gunzip
                input[2] = Channel.of([
                    [ id: 'svdb_dbs' ],
                    file(params.pipelines_testdata_base_path + 'testdata/svdb_dbs.csv', checkIfExists: true)
                ])
                input[3] = UNTAR.out.untar.map { meta, cache -> cache }
                input[4] = Channel.value('110')
                input[5] = Channel.of([
                    file(params.pipelines_testdata_base_path + 'reference/vep_plugin_files.csv', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.versions).match() }
            )
        }

    }
}
