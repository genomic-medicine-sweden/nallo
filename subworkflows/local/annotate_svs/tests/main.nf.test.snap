{
    "severus, fasta, svdb, vep_cache, '110', vep_plugins -stub": {
        "content": [
            [
                "versions.yml:md5,84c59e3a0a183b01a21620b16ee9ca4d",
                "versions.yml:md5,8b626b9e41c9e6de6edc8f94e0e75aed",
                "versions.yml:md5,cb83982d3231b281882f4ddc478d93ae"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T10:47:43.954466231"
    },
    "severus, fasta, svdb, vep_cache, '110', vep_plugins": {
        "content": [
            [
                "versions.yml:md5,84c59e3a0a183b01a21620b16ee9ca4d",
                "versions.yml:md5,8b626b9e41c9e6de6edc8f94e0e75aed",
                "versions.yml:md5,cb83982d3231b281882f4ddc478d93ae"
            ],
            "28e9be031273d5e4a94a8654da6e56ed",
            "test_svs_annotated.vcf.gz.tbi"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T10:47:32.194191605"
    }
}